### ISSUE TITLE
SFTP VFS adapter mishandles renaming directories

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2604]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2604|
|Type|defect|
|Created|2007-04-25 00:13:01Z|

The sftp adapter for vfs allows a rename of one directory to another directory to put the first directory as a sub-dir.

Steps to reproduce w/ a running SFTP server:
```
    sftp> mkdir ~user/+junk/junk
    sftp> cd ~user/+junk/junk
    sftp> mkdir directory1
    sftp> mkdir directory2
    sftp> rename directory1 directory2
    sftp> ls
    directory2
    sftp> ls directory2
    directory2/directory1
    sftp>
```
It should have failed the rename rather than putting it as a subdirectory.

Some filesystems will allow directory1 to overwrite directory2, so a better test would include adding at least one file into both directories first.

See https://bugs.launchpad.net/twisted/+bug/87934

Attachments:

* [sftp-rename-2604-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f1b/f1b0232e7a45e7f0208ceb960240cc12e75676ec/2862a64c08c684c5bd447b5400ecc6cdaa665537.diff) (1113 bytes) - added by jml on 2007-05-22 01:36:25Z - Diff of changes since last review

<details><summary>Searchable metadata</summary>

```
trac-id__2604 2604
type__defect defect
reporter__jml jml
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__vfs vfs
keywords__ 
time__1177459981000000 1177459981000000
changetime__1179825859000000 1179825859000000
version__None None
owner__ 
cc__exarkun cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml removed owner|
|-|-|

Ready for review in sftp-rename-2604

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

There is a comment in sftp that explicitly wanted this behavior: 
```
# The target node is a container.  We assume the caller means to
# move the source node into the container rather than replace
# it, and adjust newpath accordingly.
```

It seems to be like that to work like a shell (which is fine for me).

In the IETF draft, it seems it like you want. But the VFS layer has other requirement :). At least the behavior should be the same in FTP for example.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml commented|
|-|-|

The new behaviour makes conch behave like openssh sftp.

If 'rename' means one thing in SFTP and another in the shell then surely the adapter is the place to encode that difference of meaning.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

What makes me suspicious is that the change is not in conch, but in vfs.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml commented|
|-|-|

Where do you think the change should be in conch?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

Sorry, I'm not going to waste your time here because I have no idea of that should be the good behavior.

Just to explain my point: vfs provides a unique way to access a FS. For doing that, it makes some concessions on what should be the provided interface by the adapters. Thus, the current behavior of rename doesn't seem broken to me.

But, I'm not a vfs or a conch expert, so this is just my opinion, and I'll accept whatever decision is made :).

To give some review comments:
 * copyrights statements should be added
 * remove trailing whitespaces
 * the docstring of renameFile should contain "@raise" instead of "raise"

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

The interface docstring for renameFile should be updated as well.

The implementation here already has a race condition and is going to have even more serious ones when someone takes on [#10822](https://github.com/twisted/twisted/issues/10822).  This should really be fixed for real now.

It looks to me as though IFilesystemNode.rename already promises the behavior the sftp adapter wants:

```
        Renames this node to newName.  newName can be in a different
        directory.  If the destination is an existing directory, an
        error will be raised.
```

the in-memory backend even implements this behavior, but the os-fs backend fails to.  Isn't that where the real bug is?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_10"></a>|@jml removed owner|
|-|-|

Done everything except fix the race condition, for which I'll need more information / more motivation. Perhaps it should be filed as a separate ticket?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @jml|
|-|-|

The race seems inherent in the semantics offered by the VFS interfaces.  The fix is probably to change the VFS interfaces, which will break any clients attempting to rely on this semi-subtle aspect of them.

To clarify, here is the race:

```
Index: twisted/vfs/backends/osfs.py
===================================================================
--- twisted/vfs/backends/osfs.py        (revision 20128)
+++ twisted/vfs/backends/osfs.py        (working copy)
@@ -61,6 +61,10 @@
         # assumes newParent is also an OSDirectory.  Probably should politely
         # decline (rather than break with an undefined error) if it's not.
         newPath = os.path.join(newParent.realPath, pathutils.basename(newName))
+        if os.path.isdir(newPath):
+            raise ivfs.AlreadyExistsError(
+                "Can't rename %s to %s: %s already exists"
+                % (self.realPath, newPath, newPath))
```

  (here)

```
         os.rename(self.realPath, newPath)
         self.realPath = newPath
         self.name = newName
```

No extant platform which Python runs on which I am aware of can actually offer the semantics documented in the interface.  We shouldn't even try to offer this behavior, we should just do what is possible on platforms we care about.

Making this fix in as part of a separate ticket might be reasonable, but given the above, this branch is really just a bandaid, and only for launchpad (since other vfs users (if there are any) may have different assumptions) and even for launchpad, the race may allow the failure to occur (although I don't know nearly enough about how launchpad actually works to know if it is really vulnerable or not).

Glyph tells me Andy has some plans to work on VFS.  It'd be good to get his input on this.

A couple other things that I guess should be pointed out, though I don't know where exactly they lie:

  * I don't understand the actual launchpad failure, as reported on launchpad.  It says files/directories are being moved into other directories, when the expected behavior is for the rename to fail.  However, rename doesn't move files or directories in this way:

```
exarkun@charm:~/tmp$ mkdir a
exarkun@charm:~/tmp$ mkdir b
exarkun@charm:~/tmp$ touch c
exarkun@charm:~/tmp$ python
Python 2.4.3 (#2, Oct  6 2006, 07:52:30)
[GCC 4.0.3 (Ubuntu 4.0.3-1ubuntu5)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.rename('c', 'a')
Traceback (most recent call last):
  File "<stdin>", line 1, in ?
OSError: [Errno 21] Is a directory
>>> os.rename('b', 'a')
>>> os.listdir('a')
[]
>>>                                                
``` 
    so a little more insight into the actual launchpad failure would be nice.

  * OpenSSH's SFTP server (which I suppose offers the semantics launchpad wants?) uses an unatomic link/unlink pair.  This moves the race such that two files will visibly exist at some point, rather than allowing the possibility of a target file being replaced.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_12"></a>|@jml commented|
|-|-|

  I don't understand the actual launchpad failure, as reported on launchpad. It says files/directories are being moved into other directories, when the expected behavior is for the rename to fail. However, rename doesn't move files or directories in this way:

This seems to be a non sequitur. 'os.rename' doesn't move files that way, but 'rename' in SFTP while running against an osfs backend does.

 * OpenSSH SFTP offers the semantics that we want (which are the semantics defined in the draft SFTP standards, iirc).
 * I think that fixing the race condition is a separate bug to this.
 * The `if ...` check is wrong. It's ok to call os.rename and catch the failure then, as long as the exception raised plays nicely with the VFS API.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

reformatting description

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

Ah.  The move-into-directory logic was never in the backend, always in the sftp adapter.  Got it.

It looks like the conch sftp protocol implementation is slightly buggy, inasmuch as it disregards the flags sent with SSH_FXP_RENAME.  The previously implemented semantics could probably only be said to conform to the behavior defined by SSH_FXF_RENAME_NATIVE (which says "do whatever you want" ;).  A separate ticket should probably be filed against conch to add explicit support for the SSH_FXF_RENAME_OVERWRITE and SSH_FXF_RENAME_ATOMIC flags (and then OVERWRITE should be implemented or rejected with SSH_FX_OP_UNSUPPORTED).

The race condition is being fiddled around with (it was in the sftp adapter, now it's in the osfs backend), so I think fixing it in this branch is appropriate.  Changing the if to a try/except should be straightforward enough.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|z3p commented|
|-|-|

I won't get into the details of VFS and it's expectations, as I don't know that code.  However, exarkun is right in the SFTP is mishandling SSH_FXP_RENAME (judging by the implementation, the flags were added in a draft after I wrote the code).  SFTP should be updated to support this new version of the protocol: [#2619](https://github.com/twisted/twisted/issues/2619).

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_16"></a>|@jml commented|
|-|-|

OK, up for review again, `sftp-rename-2604`.

Less racy, less ambitious, more realistic.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @jml|
|-|-|

The diff to sftp.py and the tests look excellent.  A few minor things which should be easy to resolve:

  * I don't think vfs/adapters/sftp.py has been around since 2001.  The copyright statement probably shouldn't say that it has been.
  * Of the other touched files, some don't have a copyright statement and some have one that doesn't extend to 2007.  Updating these would be nice. (although I'm beginning to think we should just have a commit hook to do it)
  * The change to conch/interfaces.py doesn't look right.  The implementation of ISFTPServer in conch implements renameFile to call os.rename directly and doesn't translate to SFTPError.  I see that the sftp adapter in vfs does translate platform errors to SFTPErrors, but the sftp server in filetransfer.py (_ebStatus specifically) handles all manner of exception types (though the interface would be cleaner if it did not).  Tightening up the interface seems like an improvement, but I'd split it off into a separate ticket.
  * test-case-name needs to be on one of the first two lines of the file in order to work.  placement is preferably the first line, unless a ```#!interpreter``` needs to go there.
  * OSNode.rename has no docstring
  * AdaptFileSystemUserToISFTP's (a name only a mother could love) renameFile method has no docstring

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_19"></a>|@jml set status to `closed`|
|-|-|

(In [20351]) Raise an error when renaming a file or directory to an existing directory.

 * Author: jml
 * Fixes [#2604](https://github.com/twisted/twisted/issues/2604)
 * Reviewer: exarkun

The default SFTP server now raises an error when a user attempts to rename a
file or directory to an existing directory. This brings the behaviour in line
with the default behaviour of the OpenSSH SFTP server.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_20"></a>|@jml commented|
|-|-|

(In [20354]) Raise an error when renaming a file or directory to an existing directory.

 * Author: jml
 * Fixes [#2604](https://github.com/twisted/twisted/issues/2604)
 * Reviewer: exarkun

The default SFTP server now raises an error when a user attempts to rename a
file or directory to an existing directory. This brings the behaviour in line
with the default behaviour of the OpenSSH SFTP server.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_21"></a>|@jml commented|
|-|-|

(In [20355]) Windows failures. Refs [#2604](https://github.com/twisted/twisted/issues/2604). I can't believe I miss PQM.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_22"></a>|@jml commented|
|-|-|

(In [20357])  Raise an error when renaming a file or directory to an existing directory.

    * Author: jml
    * Fixes [#2604](https://github.com/twisted/twisted/issues/2604)
    * Reviewer: exarkun

The default SFTP server now raises an error when a user attempts to rename a
file or directory to an existing directory. This brings the behaviour in line
with the default behaviour of the OpenSSH SFTP server.

This re-commit fixes issues with Win32 support.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_23"></a>|@therve commented|
|-|-|

From ReviewProcess:

_Reverted branches are to be reviewed again before being merged._

I wouldn't insist on that if you had't break the documentation buildbot...

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_24"></a>|@therve commented|
|-|-|

(In [20358]) Revert [20357] failure in documentation slave.

#Refs [#2604](https://github.com/twisted/twisted/issues/2604).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_26"></a>|@jml removed owner<br>@jml set status to `new`|
|-|-|

Ready for review in sftp-rename-2604.

Windows and pydoc errors fixed.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_27"></a>|@therve set owner to @jml|
|-|-|

 * Unused imports in test_osfs.py
 * Please update copyrights of osfs.py and test_osfs.py

Then you can merge, thanks.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_28"></a>|@jml set status to `closed`|
|-|-|

(In [20365]) Raise an error when renaming a file or directory to an existing directory.

    * Author: jml
    * Fixes [#2604](https://github.com/twisted/twisted/issues/2604)
    * Reviewers: exarkun, therve

The default SFTP server now raises an error when a user attempts to rename a
file or directory to an existing directory. This brings the behaviour in line
with the default behaviour of the OpenSSH SFTP server.

This re-commit fixes issues with Win32 support and corrects an epytext error.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
